### PR TITLE
Security hardening

### PR DESCRIPTION
**Pull request by Dude4Linux**
References:
[Guide to Deploying Diffie-Hellman for TLS](https://weakdh.org/sysadmin.html)
[Configuring Apache, Nginx, and OpenSSL for Forward Secrecy](https://community.qualys.com/blogs/securitylabs/2013/08/05/configuring-apache-nginx-and-openssl-for-forward-secrecy)
[Strong SSL Security on Apache2](https://raymii.org/s/tutorials/Strong_SSL_Security_On_Apache2.html)
[Logjam: the latest TLS vulnerability explained](https://blog.cloudflare.com/logjam-the-latest-tls-vulnerability-explained/)
[BREACH mitigation](https://community.qualys.com/message/20404#20404)
[Configuring Apache for Perfect Forward Secrecy](http://www.lorrin.org/blog/2013/07/03/configuring-apache-for-perfect-forward-secrecy/)
[Eveho Blog - SSL/TLS, ciphers, perfect forward secrecy and Tomcat](https://blog.eveoh.nl/2014/02/tls-ssl-ciphers-pfs-tomcat/)
Testing:
[testssl.sh](https://github.com/drwetter/testssl.sh) - use git clone to install on your scanning host
nmap - $ sudo nmap -p 10000-20000 -sSV -T4 <hostname>
[OpenVAS](http://www.openvas.org/) - available on [Kali Linux](https://www.kali.org/)

### CODE REVIEW

**Comment by Dude4Linux** [comment]
I still have some changes and a little cleanup to do, but I wanted to get this out for review and discussion.

**Comment by Dude4Linux** [comment]
Don't you just hate when you've completed testing and everything seems to be working, you step back and review your code, then immediately see a better way you could have done it.

**Comment by Dude4Linux** [comment]
Decided it was better to consolidate the security updates for all applications into one script that runs on all appliances rather than a separate script for each app.  Script must test for the presence of an app's config file before applying changes.

**Comment by Dude4Linux** [comment]
I pushed the turnkey-make-ssl-cert and associated turnkey.cnf files for evaluation and discussion.
WARNING:  Do not run this script on a TurnKey appliance with the -d (--default) option.  It will replace your cert.pem file without asking.  Been there, Done that!

**Comment by Dude4Linux** [comment]
I think I have finally managed to lick the stunnel config issues by dividing the processing into two routines, one for options and the other for params.  A single routine could not handle the logic needed to process both.  I tried to make both routines idempotent and as general purpose as possible so they may be used with other file formats, apache.conf, ini files, etc.  Have not yet tried them with other config yet.

**Comment by Dude4Linux** [comment]
After moving the security script back to common/conf, I'm finding that the remove_ssl_params routine is not effective.  The appliance overlays e.g. drupal7 have not been applied when the script runs.  Since we still need to restart services during firstboot, I propose moving the 'remove_ssl_params' routine to firstboot.d/50ssl_cleanup.  I can put in a general grep for /etc/ssl/certs/cert.pem and change it to /etc/ssl/private/cert.pem.

File: overlays/turnkey.d/security/etc/ssl/turnkey.cnf [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#
+# SSL template for TurnKey GNU/Linux appliances
+# Based on the SSLeay example configuration file.
+#
+
+RANDFILE                = /dev/urandom
```
Comment (JedMeister) [inline overlays/turnkey.d/security/etc/ssl/turnkey.cnf]: IIRC ideally /dev/random should be used instead. /dev/urandom is easier/nicer in some respects as it's non-blocking, but AFAIK /dev/random provides superior pseudo random numbers...

**Comment by Dude4Linux** [comment]
This latest change resulted in my first successful build of the Drupal7 appliance without needing any tweaks to run.  There is still a failure to start apache on firstboot because it is started using defaults before any firstboot.d script is run.  So to summarize, common/conf/turnkey.d is too early, and firstboot.d is too late.  There really isn't any alternative to tracking down and fixing all references to ssl in the appliance overlays.  Sorry @JedMeister  :)

**Comment by Dude4Linux** [comment]
Not sure what to think about the last two commits.  I was doing some code cleanup, removing what I thought was unnecessary code.  After rebuilding the drupal7 appliance and rescanning, I spotted the following
```
 Server banner                MiniServ/1.760
 Application banner           --
 Cookie(s)                    1 issued: NOT secure, 1/1 HttpOnly
 Security headers             X-Frame-Options: SAMEORIGIN
                              Content-Security-Policy: script-src 'self' 'unsafe-inline'; frame-src 'self'
```
Comparing against the previous scan, I thought the change was due to removing the ssl params from webmins config.  However, reverting the change did not seem to help.
```
 Server banner                MiniServ/1.740
 Application banner           --
 Cookie(s)                    1 issued: 1/1 secure, 1/1 HttpOnly
 Security headers             X-Frame-Options: SAMEORIGIN
                              Content-Security-Policy: script-src 'self' 'unsafe-inline'; frame-src 'self'
```
Now I see that upstream has changed the webmin version so that is likely the cause, so these last two commits are probably unnecessary.

**Comment by Dude4Linux** [comment]
@JedMeister I tried to submit a change to the openvpn appliance to configure lighty to use the ssl parms from /etc/lighttpd/ssl-parms.conf created during build.  I hadn't forked the appliance so I wasn't able to push a new branch.  Anyway here is a patch for openvpn.
```
diff --git a/conf.d/main b/conf.d/main
index d6cc496..307a4a8 100755
--- a/conf.d/main
+++ b/conf.d/main
@@ -19,7 +19,6 @@ sed -i "s|^CRL=.*|CRL=\"crl.jail/crl.pem\"|" $EASY_RSA/revoke-full
 # webserver related
 mkdir -p /var/www/openvpn/logs
 chown www-data:www-data /var/www/openvpn/logs
-lighty-disable-mod ssl
 
 mkdir -p $HTDOCS/{css,images}
 cp /var/www/js/ui.* $HTDOCS/js/
diff --git a/overlay/etc/lighttpd/lighttpd.conf b/overlay/etc/lighttpd/lighttpd.conf
index ff05fa8..afb1805 100644
--- a/overlay/etc/lighttpd/lighttpd.conf
+++ b/overlay/etc/lighttpd/lighttpd.conf
@@ -42,9 +42,7 @@ $SERVER["socket"] == ":80" {
 
 $SERVER["socket"] == ":443" {
     ssl.engine = "enable"
-    ssl.pemfile = "/etc/ssl/certs/cert.pem"
-    ssl.cipher-list = "ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-SHA256::ECDHE-RSA-AES256-SHA256:ECDHE-RSA-AES128:ECDHE-RSA-AES256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA:RC4-SHA"
-    ssl.honor-cipher-order = "enable"
+    include "ssl-params.conf"
 
     server.document-root = "/var/www/openvpn/htdocs/"
     accesslog.filename = "/var/www/openvpn/logs/access.log"
```

**Comment by Dude4Linux** [comment]
@JedMeister I happened to notice the following during testing and think it needs to be investigated.
```
# ll /etc/ssl/certs | grep ' cert.pem'
lrwxrwxrwx 1 root root      8 Aug 13 12:08 54a05ec5.0 -> cert.pem
lrwxrwxrwx 1 root root      8 Aug 13 12:08 b0729537.0 -> cert.pem
lrwxrwxrwx 1 root root     41 Aug 13 12:08 cert.pem -> /usr/local/share/ca-certificates/cert.crt
```
Note the two softlinks pointing to cert.pem.  IMO there should only be one.  I think the first one is created as a side effect of moving the default cert.crt into /usr/local/share/ca-certificates/cert.crt.  Originally I had placed it into /etc/ssl/private, but when I saw that it wasn't being replaced by turnkey-make-ssl-cert, I moved it to /usr/local/share/ca-certificates/cert.crt so it would be overwritten.  I see that update-ca-certificates is being run during build when the ca-certificates package is installed.  I think that is what generates the first softlink.  What I can't explain is why that link wasn't removed when the cert and key were generated.  I'm going to try just removing the default key and crt leaving only the cert.pem.

**Comment by Dude4Linux** [comment]
The last commit did not change the double links to -> cert.pem
but I think it was good to remove the files.  Only the default cert.pem is required to complete the build.
The cause of the two softlinks is still a mystery.  Maybe it's legit.

**Comment by JedMeister** [comment]
Thanks for the further info @Dude4Linux 

TBH the volume of work you've done is mammoth! I will look into the multiple symlinks you mention.

I will laos make sure that we move the cert location on all other apps...

**Comment by Dude4Linux** [comment]
I found that the recommendation from Raymii.org to disable ssl compression on Lighty is not needed
```
ssl.use-compression = "disable"
```
because ssl compression is not used unless mod_compress is enabled and Lighty complains in the error.log when this parameter is used.

**Comment by Dude4Linux** [comment]
After reviewing the recommended configs at Cipherli.st from Raymii (I think), I see there is one more lighttpd param that could be set.
```
setenv.add-response-header = (
    "Strict-Transport-Security" => "max-age=63072000; includeSubdomains; preload",
    "X-Frame-Options" => "DENY",
    "X-Content-Type-Options" => "nosniff"
)
```
I won't have time to test this so if some wants to take a shot, be my guest.

**Comment by JedMeister** [comment]
@qrntz - @Dude4Linux & I have discussed this (via email?) and I thought you were involved (but obviously not). Regardless it probably should have been stated explicitly in public... 

I intend to fix the appliances individually. I have also implemented a lighty (shared/common) overlay to remove any need for firstboot (in relation to lighty anyway) and testing seems good. I have made a few adjustments for Apache too (simplified by moving some stuff from conf to the overlay) but need to do more testing... Also nginx may also benefit from a shared/common overlay too?

**Comment by qrntz** [comment]
@JedMeister, I see. Sounds good then.

I'll attend to nginx.

**Comment by JedMeister** [comment]
Seeing as this seems to be the place that we're discussing all this; @Dude4Linux & @qrntz FYI:

My current common branch for this work is [here](https://github.com/JedMeister/common/tree/ssl-hardening). It is a new branch I've made today which consists of John's work to date, the tweaking that I've done since I started and some commits from Anton too (namely stuff John had already merged; plus Tomcat stuff)...:

Here is where I'm up to:
- moved stunnel.conf from conf script to pre-existing overlay
   - left cipher setting in script
- moved lighty to (new) common overlay (ssl-params.conf as per what John's script was creating)
    - added `setenv.add-response-header` as per John's comment (above)
    - left cipher setting in script
- adjusted all (individual) lighty apps to use new common overlay (inc shared ssl settings)
- moved apache conf from conf script to overlays (ssl.conf already existed; security.conf is new)
    - left cipher setting in script
- adjusted (individual) apache apps to remove ssl cert altogether (as discussed elsewhere; it's now included in default conf overlay)
 - moved tomcat conf from conf file to overlay
    - left cipher setting in script
- fixed webmin (I think that someone accidentally broke it at some point...)

So far I have only been building to root.patched and not running the firstboot script as IMO we need to get rid of it anyway. So my next step will be to remove it altogether but it's unlikely to happen today...

All this SSL stuff is very close to being done; just need to finalise nginx, remove the firstboot script and do some more testing...

Have a look here if you're interested:
https://github.com/turnkeylinux/common/compare/master...JedMeister:ssl-hardening?expand=1

**Comment by JedMeister** [comment]
Almost forgot to mention I haven't been seeing the multiple cert symlinks in vanilla root.patched (in fact there aren't any in vanilla root.patched...) so it must be being created by one of the inithooks or something one of the inithooks is calling...

**Comment by qrntz** [comment]
@JedMeister, great job! :+1:

**Comment by Dude4Linux** [comment]
@JedMeister and @qrntz Let me second that.  Great job for both of you!  Thanks for the feedback and tips!  Looking forward to a solid and secure 14.0 release when I get back.  Now I really gotta get packing.

@JedMeister re: the two symlinks.  I confirmed that no symlinks were present when using fab-chroot build/root.patched, and that two symlinks exist after firstboot.d scripts are run.  I thought about adding a check right before turnkey-make-ssl-cert is run in 15regen-sslcert and removing and links found. I've removed all other places where update-ca-certificates was run except the one in turnkey-make-ssl-cert.  Sorry I don't have time to check this any further. Gotta run.

**Comment by DocCyblade** [comment]
@JedMeister asked me to do some testing. Testing on the below apps. Using this post to document my findings. I'll edit this as my testing unfolds. I'll capture all output from testssl script as nameOfApp-test1, nameOfApp-test2, and nameOfApp-test3 and attach them to this as well.  

### Apps Tested
- [ ] Openvpn
- [ ] Openldap
- [ ] Mongodb
- [ ] lighty itself (lighttpd-fastcgi-etc)
- [ ] mySQL
- [ ] PostgreSQL




------
### Apps Tested Details
- [ ] Test Openvpn
    - [X] Basic Build to CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] SSL Test 2 (after init-hooks) in CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] Build ISO
    - [ ] Install ISO on VM
    - [ ] SSL Test 3 in VM

- [ ] Test Openldap
    - [X] Basic Build to CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] SSL Test 2 (after init-hooks) in CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] Build ISO
    - [ ] Install ISO on VM
    - [ ] SSL Test 3 in VM

- [ ] Test Mongodb
    - [X] Basic Build to CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] SSL Test 2 (after init-hooks) in CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] Build ISO
    - [ ] Install ISO on VM
    - [ ] SSL Test 3 in VM

- [ ] Test lighty itself (lighttpd-fastcgi-etc)
    - [X] Basic Build to CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] SSL Test 2 (after init-hooks) in CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] Build ISO
    - [ ] Install ISO on VM
    - [ ] SSL Test 3 in VM

- [ ] Tesy mySQL
    - [X] Basic Build to CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] SSL Test 2 (after init-hooks) in CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] Build ISO
    - [ ] Install ISO on VM
    - [ ] SSL Test 3 in VM

- [ ] Test PostgreSQL
    - [X] Basic Build to CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] SSL Test 2 (after init-hooks) in CHROOT
    - [ ] SSL Test 1 in CHROOT
    - [ ] Build ISO
    - [ ] Install ISO on VM
    - [ ] SSL Test 3 in VM


-----
Testing being done with common branch and the ssl-harding branches of each. Snap shot of the commits working with
```
root@tkldev-l fab/common# git branch -v
* ssl-hardening 01fcb46 moved tomcat conf to overlay file & fixed webmin

root@tkldev-l products/lighttpd-php-fastcgi# git branch -v
* ssl-hardening-2 53b62fd enabled shared lighty overlay & edited default site

root@tkldev-l products/mongodb# git branch -v
* ssl-hardening 7326fcd use common overlay & adjust lighty site

root@tkldev-l products/mysql# git branch -v
* ssl-hardening a0bd21c use common overlay & adjust lighty site

root@tkldev-l products/openldap# git branch -v
* ssl-hardening 2c596c8 use common overlay & adjust lighty site

root@tkldev-l products/openvpn# git branch -v
* ssl-hardening a229559 use common overlay & adjust lighty site

root@tkldev-l products/postgresql# git branch -v
* ssl-hardening 5543d18 use common overlay & adjust lighty site

```

Testing methodology:
- build app to root.patched
- fab-chroot enter; start (specific) services and prelim test
   - every appliance tested:
       - start webserver
           - starts ok?
       - test ports 443 & 12322 SSL with [testssl.sh](https://github.com/drwetter/testssl.sh)
           - complies with previous tests? (run diff against attached
recent ones)
           against it)
   - at least once and randomly after:
       - start stunnel4; webmin & shellinabox
           - starts ok?
       - test ports 12320 & 12321 SSL with [testssl.sh](https://github.com/drwetter/testssl.sh)
- run init hooks (turnkey-init)
   - re-run  [testssl.sh](https://github.com/drwetter/testssl.sh) tests
- build all the way to ISO
   - do this step at least once and randomly there-after
   - install ISO in VM
   - run though first boot
   - test using  [testssl.sh](https://github.com/drwetter/testssl.sh) (as per above)

**Comment by JedMeister** [comment]
@DocCyblade - Legend! :+1:

**Comment by JedMeister** [comment]
Happy travels @Dude4Linux  :smile:

**Comment by DocCyblade** [comment]
I spent most of my time creating a script to test with :-) See my post for details. I will be back at it after 4pm today. If the script works as planned I can test pretty quickly with results captured in github

**Comment by DocCyblade** [comment]
Testing has started again, I should have everything tested on the list in 3-4 hours (11pm EST) Check above post for details (https://github.com/turnkeylinux/common/pull/42#issuecomment-132284465)

**Comment by JedMeister** [comment]
@DocCyblade - Great work mate! :+1: Looks like it could be a useful framework for testing other appliances too... We'd just need to add some other services (for other apps)...

**Comment by DocCyblade** [comment]
First wave is done, going to test the apache tomorrow

**Comment by JedMeister** [comment]
Thanks again @DocCyblade! :+1:

**Comment by DocCyblade** [comment]
Since the last two failed I added yiiframework, code igniter and jenkins, all based on LAMP. I'll be testing them shortly

**Comment by DocCyblade** [comment]
I started testing yii, and got errors too. I am wondering if it is not a caching issue I ran into before ([See here for the details](https://www.turnkeylinux.org/forum/support/20150701/tkldev-builds-very-slow-build#comment-23419))

I cleared the cache and yii is now looks like it's building now. I will circle back around and retest the above failed ones with a cleared cache

**Comment by DocCyblade** [comment]
The other two are still not building. I tried to build code and get below error constantly.  Posting here as an FYI.


Error below
```
Processing triggers for systemd (215-17+deb8u1) ...
Errors were encountered while processing:
 /var/cache/apt/archives/mysql-server-5.5_5.5.44-0+deb8u1_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
Traceback (most recent call last):
  File "/usr/lib/fab/wrapper.py", line 34, in <module>
    CliWrapper.main()
  File "/usr/lib/python2.7/dist-packages/pyproject.py", line 388, in main
    commands.run(name, args)
  File "/usr/lib/python2.7/dist-packages/pyproject.py", line 261, in run
    command.module.main()
  File "/usr/lib/fab/pylib/cmd_install.py", line 129, in main
    installer.install(packages, ignore_errors)
  File "/usr/lib/fab/pylib/installer.py", line 267, in install
    self._install(packages, ignore_errors)
  File "/usr/lib/fab/pylib/installer.py", line 137, in _install
    self.chroot.system("apt-get", *(args + packages))
  File "/usr/lib/python2.7/dist-packages/chroot.py", line 80, in system
    executil.system(*self._prepare_command(*command))
  File "/usr/lib/python2.7/dist-packages/executil.py", line 56, in system
    raise ExecError(command, exitcode)
executil.ExecError: non-zero exitcode (100) for command: chroot '/turnkey/fab/products/codeigniter/build/root.build'
```

**Comment by JedMeister** [comment]
@DocCyblade 
>````
Errors were encountered while processing:
 /var/cache/apt/archives/mysql-server-5.5_5.5.44-0+deb8u1_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)````

IME that is caused by MySQL running somewhere on the tkldev host (probably within one of the chroots - often from a previous failed build). It's because as part of the install of the package it starts the server and obviously if it's already running it can't start (hence the failure). Do a `ps aux | grep mysql` (not in a chroot) to see if I'm right! :smile:

**Comment by DocCyblade** [comment]
@JedMeister - You are too smart!  You were correct. I knew it must have been me since anything with mysql was failing to build. I ended up rebooting and running my "clean_up" script. (It resets TKLDev to factory out of the box) I'll post it or add it to the TKLDev tool chain when I work out the bugs. That worked too

Once I finish the testing, I'll take a look at the other two bug issues today

**Comment by DocCyblade** [comment]
@JedMeister all apps tested. I'll start work on Symfony

**Comment by JedMeister** [comment]
:+1:

**Pull request merged**